awx/main/dispatch/worker/callback.py

```diff
@@ -66,8 +72,19 @@ def read(self, queue):
             logger.exception("failed to decode JSON message from redis")
         finally:
             self.record_statistics()
+            self.record_read_metrics()
+
         return {'event': 'FLUSH'}
 
+    def record_read_metrics(self):
+        if self.queue_pop == 0:
+            return
+        if self.subsystem_metrics.should_pipe_execute() is True:
```

Comment:
I think there have been some changes to the code since I last looked. Before, I was observing negative stats for the events in memory statistic. I think it can still happen here if the decrement operation below always "takes" the pipeline execution for the time window. The addition here could get starved while the system is busy, resulting in a negative number.

Reply:
from my understanding, "read" is always called before "flush" since it runs in the same thread/loop. Therefore, the decrement in "flush" will always occur after the increment in "read". The reason we were getting negative values before was that we were delaying the increment operation in read. And during that delay, it was possible to decrement in flush.